Gems/AtomTressFX/Code/Passes/HairPPLLRasterPass.cpp

```diff
@@ -0,0 +1,315 @@
+                pipelineStateDescriptor.m_inputStreamLayout.SetTopology(AZ::RHI::PrimitiveTopology::TriangleList);
+                pipelineStateDescriptor.m_inputStreamLayout.Finalize();
+
+                m_pipelineState = m_shader->AcquirePipelineState(pipelineStateDescriptor);
+                if (!m_pipelineState)
+                {
+                    AZ_Error("Hair Gem", false, "Pipeline state could not be acquired");
+                    return false;
+                }
+
+                RPI::ShaderReloadNotificationBus::Handler::BusConnect(shaderAsset.GetId());
+
+                m_initialized = true;
+                return true;
+            }
+
+            void HairPPLLRasterPass::SchedulePacketBuild(HairRenderObject* hairObject)
+            {
+                m_newRenderObjects.insert(hairObject);
+            }
+
+            bool HairPPLLRasterPass::BuildDrawPacket(HairRenderObject* hairObject)
+            {
+                if (!m_initialized)
+                {
+                    return false;
+                }
+
+                RHI::DrawPacketBuilder::DrawRequest drawRequest;
+                drawRequest.m_listTag = m_drawListTag;
+                drawRequest.m_pipelineState = m_pipelineState;
+//                drawRequest.m_streamBufferViews =  // no explicit vertex buffer.  shader is using the srg buffers
+                drawRequest.m_stencilRef = 0;
+                drawRequest.m_sortKey = 0;
+
+                // Seems that the PerView and PerScene are gathered through RenderPass::CollectSrgs()
+                // The PerPass is gathered through the RasterPass::m_shaderResourceGroup
+                AZStd::lock_guard<AZStd::mutex> lock(m_mutex);
+
+                return hairObject->BuildPPLLDrawPacket(drawRequest);
+            }
+
+            bool HairPPLLRasterPass::AddDrawPacket(HairRenderObject* hairObject)
+            {
+                const RHI::DrawPacket* drawPacket = hairObject->GetFillDrawPacket();
+                if (!drawPacket)
+                {   // might not be an error - the object might have just been added and the DrawPacket is
+                    // scheduled to be built when the render frame begins
+                    AZ_Warning("Hair Gem", !m_newRenderObjects.empty(), "HairPPLLRasterPass - DrawPacket wasn't built");
+                    return false;
+                }
+
+                if (!m_currentView &&
+                    (!(m_currentView = GetView()) || !m_currentView->HasDrawListTag(m_drawListTag)))
+                {
+                    m_currentView = nullptr;    // set it to nullptr to prevent further attempts this frame
+                    AZ_Warning("Hair Gem", false, "HairPPLLRasterPass failed to acquire or match the DrawListTag - check that your pass and shader tag name match");
+                    return false;
+                }
+
+                m_currentView->AddDrawPacket(drawPacket);
+                return true;
```

Comment:
Why not do this from the Feature Processor instead of going through the Pass here?

Reply:
It is really a matter of choice, but in this case the FP have many passes unlike most FPs and each pass is different (multiple compute, regular render (this one), and full screen render and each works differently.
I rather keep the FP more pure and not involved in each of the passes' specific method of operation relevant only to its type.
In this case, only the render pass cares about the view and tags so it makes sense to let it handle it and not the FP.